### PR TITLE
Adding a new RepositoryTrait to make service repos easy

### PR DESCRIPTION
**Pull request by weaverryan**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | none
| License       | MIT
| Doc PR        | TODO

See https://github.com/doctrine/DoctrineBundle/pull/725#issuecomment-339421794

The goal is to allow people to autowire & autoregister their repositories. Instead of actually trying to make the true repositories into services in some automatic way (an approach which has failed due to some technical reasons), this proposal simplifies creating your own services.

The new`RepositoryTrait` makes creating a new service that decorates your repository is very easy. And the PR (as you can see) is about 50 times simpler :).

Usage:

* You are still allowed to map a `repositoryClass` on your entity, but generally, you will not (create a service instead).

* Create a repository service:

```php
<?php

namespace App\Repository;

use Doctrine\Bundle\DoctrineBundle\Repository\RepositoryTrait;
use Symfony\Bridge\Doctrine\RegistryInterface;
use App\Entity\CoolStuff;

class CoolStuffRepository
{
    use RepositoryTrait;

    protected function getClassName()
    {
        return CoolStuff::class;
    }

    // add custom methods here
    // you have access to `$this->getEntityManager()` and `$this->getRepository()` methods
    // (in addition to createQueryBuilder(), and all the other normal methods)
    // this works because of autowiring & the @required annotation above setEntityManager()
}
```

The `RepositoryTrait` gives you all the existing methods on `EntityRepository`, like `createQueryBuilder()` and `find**` (but no magic methods). If the user wants to make their class implement the `ObjectRepository` interface, that's totally up to them. You can also ignore our trait and create your services all on your own.

To use it, just DI your class like normal. 

For entities with NO custom repository, users can still use `$em->getRepository()`. But when they need custom methods, they would create a custom service and start using it.

Thanks!

### CODE REVIEW

File: src/Symfony/Bridge/Doctrine/Repository/RepositoryTrait.php [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+ * A helpful trait when creating your own repository.
+ *
+ * @author Ryan Weaver <ryan@knpuniversity.com>
+ */
+trait RepositoryTrait
+{
+    /**
+     * @return EntityManagerInterface
+     */
+    abstract protected function getEntityManager();
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Doctrine/Repository/RepositoryTrait.php]: what about making this method concrete but adding a corresponding setEntityManager method in the trait, with the `@required` annotation?

Comment (nicolas-grekas) [reply src/Symfony/Bridge/Doctrine/Repository/RepositoryTrait.php]: you could in fact remove the getter method and provide a private property instead

File: src/Symfony/Bridge/Doctrine/Repository/RepositoryTrait.php [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+     */
+    public function matching(Criteria $criteria)
+    {
+        return $this->getRepository()->matching($criteria);
+    }
+
+    /**
+     * @return EntityManagerInterface
+     */
+    protected function getEntityManager()
```
Comment (Tobion) [inline src/Symfony/Bridge/Doctrine/Repository/RepositoryTrait.php]: getEntityManager and getRepository should be marked as final so we can return types in sf 4

File: src/Symfony/Bridge/Doctrine/Repository/RepositoryTrait.php [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+use Doctrine\Common\Collections\Criteria;
+use Doctrine\ORM\EntityManagerInterface;
+use Doctrine\ORM\EntityRepository;
+use Doctrine\ORM\NativeQuery;
+use Doctrine\ORM\Query;
+use Doctrine\ORM\Query\ResultSetMappingBuilder;
+use Doctrine\ORM\QueryBuilder;
+
+/**
+ * A helpful trait when creating your own repository.
```
Comment (Tobion) [inline src/Symfony/Bridge/Doctrine/Repository/RepositoryTrait.php]: you should point out that you only need this trait if you want to expose the same set of methods as doctrine entityrepository.

**Comment by ro0NL** [comment]
> You are still allowed to map a repositoryClass on your entity

But then you need to implement the builting entity repository again.. :confused:  Given `public function getRepository($entityName): \Doctrine\ORM\EntityRepository`, thus defeating this feature.

You would have 2 repo's then? The builtin version + one using the trait (which internally uses the builtin version again).

> But when they need custom methods, they would create a custom service and start using it.

Or a custom `EntityRepository`...

IMHO repo's as a service can live fully in userland. Im not convinced of the added value of this trait. I might not need everything it's providing now, and if so id extend `EntityRepository` anyway ;-)

**Comment by weaverryan** [comment]
> But then you need to implement the builting entity repository again.. 😕 Given public function getRepository($entityName): \Doctrine\ORM\EntityRepository, thus defeating this feature.

You won't need to actually implement the EntityRepository (create a class or configure anything). Your service will use the EntityRepository internally, but that's invisible to the user.

> IMHO repo's as a service can live fully in userland. Im not convinced of the added value of this trait. I might not need everything it's providing now, and if so id extend EntityRepository anyway ;-)

Unfortunately, you can't extend `EntityRepository` from your own service... well... not easily :)

I just pushed changes thanks to Tobion. But, I'm waiting on some information from @beberlei (https://github.com/doctrine/DoctrineBundle/pull/725#issuecomment-339666270) about multiple entity managers. We may be able to inject the Registry instead, and use it to dynamically select the correct EM.

File: src/Symfony/Bridge/Doctrine/Repository/RepositoryTrait.php [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+     */
+    public function findOneBy(array $criteria, array $orderBy = null)
+    {
+        return $this->getRepository()->findOneBy($criteria, $orderBy);
+    }
+
+    /**
+     * @see EntityRepository::matching()
+     *
+     * @param \Doctrine\Common\Collections\Criteria $criteria
```
Comment (ro0NL) [inline src/Symfony/Bridge/Doctrine/Repository/RepositoryTrait.php]: `@param Criteria $criteria`

File: src/Symfony/Bridge/Doctrine/Repository/RepositoryTrait.php [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+     *
+     * @return \Doctrine\Common\Collections\Collection
```
Comment (ro0NL) [inline src/Symfony/Bridge/Doctrine/Repository/RepositoryTrait.php]: `@return Collection`

**Review by ro0NL — COMMENTED** [review]
Changed my vote to :+1: wouldnt use this myself... but i understand we need to do something to promote repos as a service quick&easy.

The concern remains where people configuring `repositoryClass` today (with a `CustomEntityRepository`). They need to be aware how this feature reflects to that (to not end up with 2 repos).

**Comment by weaverryan** [comment]
Don't merge this yet: DoctrineBundle will likely get support again for PHP 5.5. And I'm also still looking into an alternate approach (which would be done in that bundle). But the plan would be to merge this *or* that other approach.. which I'm working on :).

**Comment by fabpot** [comment]
closing in favor of doctrine/DoctrineBundle#727

**Pull request closed without merging**